src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs

```diff
@@ -537,53 +549,52 @@ private bool TryGetNodeWithEditableSignatureOrAttributes(Location location, Solu
                 }
             }
 
-            // 7. Add the remaining arguments. These will already have names or be params arguments, but may have been removed.
+            // 6. Add the remaining arguments. These will already have names or be params arguments, but may have been removed.
             var removedParams = updatedSignature.OriginalConfiguration.ParamsParameter != null && updatedSignature.UpdatedConfiguration.ParamsParameter == null;
-            for (var i = declarationParametersToPermute.Count; i < arguments.Count; i++)
+            for (var i = declarationParametersToPermute.Length; i < arguments.Length; i++)
             {
-                if (!arguments[i].IsNamed && removedParams && i >= updatedSignature.UpdatedConfiguration.ToListOfParameters().Count)
+                if (!arguments[i].IsNamed && removedParams && i >= updatedSignature.UpdatedConfiguration.ToListOfParameters().Length)
                 {
                     break;
                 }
 
-                if (!arguments[i].IsNamed || updatedSignature.UpdatedConfiguration.ToListOfParameters().Any(p => p.Name == arguments[i].GetName()))
+                if (!arguments[i].IsNamed || updatedSignature.UpdatedConfiguration.ToListOfParameters().Any<Parameter>(p => p.Name == arguments[i].GetName()))
                 {
                     newArguments.Add(arguments[i]);
                 }
             }
 
-            return newArguments;
+            return newArguments.ToImmutableAndFree();
         }
 
         private static SignatureChange CreateCompensatingSignatureChange(ISymbol declarationSymbol, SignatureChange updatedSignature)
         {
-            if (declarationSymbol.GetParameters().Length > updatedSignature.OriginalConfiguration.ToListOfParameters().Count)
+            if (declarationSymbol.GetParameters().Length > updatedSignature.OriginalConfiguration.ToListOfParameters().Length)
             {
                 var originalConfigurationParameters = updatedSignature.OriginalConfiguration.ToListOfParameters();
                 var updatedConfigurationParameters = updatedSignature.UpdatedConfiguration.ToListOfParameters();
 
                 var realParameters = declarationSymbol.GetParameters();
+                var bonusParameters = realParameters.Skip(originalConfigurationParameters.Length);
 
-                var bonusParameters = realParameters.Skip(originalConfigurationParameters.Count);
+                var compensatedOriginalConfigurationParameters = originalConfigurationParameters.AddRange(bonusParameters.Select(p => new ExistingParameter(p)));
+                var compensatedUpdatedConfigurationParameters = updatedConfigurationParameters.AddRange(bonusParameters.Select(p => new ExistingParameter(p)));
```

Comment:
what does 'compensated' mean?

Reply:
Changed the symbol names here and added a doc comment. It's basically when you're reordering a delegate parameter list with 2 parameters, but you have to worry about BeginInvoke calls with 4 arguments. This adds those extra 2 on.